### DISCUSSION TITLE
Is it possible for region variables to escape through a closure?

### DISCUSSION THREAD

**Discussion by mbarbar** [flix/flix#7143] [General]
```
def main(): Unit \ IO =
    let x = region rc {
        let s = MutSet.new(rc);
        MutSet.add!(123, s);
        MutSet.add!(456, s);
        () -> { MutSet.forEach(println, s) }
    };
    x()
```

This program attempts to operate on `s`, which belongs to `rc`, outside `rc`, and compiles and produces
```
123
456
```

Is this a bug or intended? I think either could be justified, though

> Regions and all memory associated with them cannot outlive their lexical scope.

shows that there is at the least a contradiction between the implementation and the documentation currently.

Upon search I came across https://github.com/flix/flix/issues/6142, but I didn't quite understand what's being discussed there.

**Comment by magnus-madsen** [comment]
This is a (known) bug/limitation. Memory in regions should __never__ outlive their lexical scope.

However, this is currently "allowed" because `IO` represent any arbitrary effect.

I believe this very fresh PR #7114 fixes your example.

**Comment by magnus-madsen** [comment]
TLDR: This is just a corner case of our implementation which will be fixed soon.

**Comment by magnus-madsen** [comment]
```
-- Type Error -------------------------------------------------- foo.flix

>> The region variable 'rc' escapes its scope.

2 |>     let x = region rc {
3 |>         let s = MutSet.new(rc);
4 |>         MutSet.add!(123, s);
5 |>         MutSet.add!(456, s);
6 |>         () -> { MutSet.forEach(println, s) }
7 |>     };

```
